Don't smoosh hostname and domainname in API

thockin opened the pull request:
Fixes #14282 

I think this is a better solution than #20196.  Rather than make the CLI split `--hostname`, let it happen at the backend.  This preserve's the user's intentions all the way through.  No API change needed.  No behavioral change from the CLI.  The results of `docker inspect` will be different for `--hostname=foo.bar.com` containers, but the user-visible behavior is the same.  The only behavioral change is for API users that specify both `hostname` and `domainname`, in which case the container's hostname will just be the `hostname` field.  Otherwise it behaves exactly the same.

Tests and docs are not updated yet, pending someone saying they don't hate this.  And yes, it obviously has to go into libnetwork, not here.  I'll do all that if folks say this is OK.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14282-hostname-domainname-v2" git@github.com:thockin/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

estesp commented:
Haven't had a chance to play with this, but it looks like a promising fix to the originally thorny problem of figuring out the intent vs. distro differences that brought up the original bug.

When you say `docker inspect` is changed, I assume you mean it will just show the entire FQDN as `Hostname` with a blank `Domainname` for the CLI usecase of `--host=hostname.domainname`, which would have been shown separately due to the splitting logic being at the "front-end" prior to this PR.

I think the only downside is that for non-API uses with a FQDN in `--host`, this PR doesn't fix the max length problem for the `SetHostname` syscall. I remember one of the difficulties when I worked on the original "fix" is that even though the Docker container config has the concept of a separate hostname/domainname, **libcontainer** [does not](https://github.com/opencontainers/runc/blob/master/libcontainer/configs/config.go#L108-L109), and the passthrough is the somewhat clunky env. variable `HOSTNAME` set into the process config buffer, which the execdriver [then uses](https://github.com/docker/docker/blob/master/daemon/execdriver/driver_unix.go#L140) to set up the libcontainer variant of the container's config. However, at least this allows for proper use via the API and with the libnetwork changes, gives you a more appropriate setup for differentiating hostname and domainname, and more importantly fixes the hostname length max issue for that use case.

thockin commented:
Yes, your assessment of the This doesn't fix the length issue

On Wed, Feb 10, 2016 at 11:51 AM, Phil Estes <notifications@github.com>
wrote:

> Haven't had a chance to play with this, but it looks like a promising fix
> to the originally thorny problem of figuring out the intent vs. distro
> differences that brought up the original bug.
>
> When you say docker inspect is changed, I assume you mean it will just
> show the entire FQDN as Hostname with a blank Domainname for the CLI
> usecase of --host=hostname.domainname, which would have been shown
> separately due to the splitting logic being at the "front-end" prior to
> this PR.
>
> I think the only downside is that for non-API uses with a FQDN in --host,
> this PR doesn't fix the max length problem for the SetHostname syscall. I
> remember one of the difficulties when I worked on the original "fix" is
> that even though the Docker container config has the concept of a separate
> hostname/domainname, *libcontainer* does not
> <https://github.com/opencontainers/runc/blob/master/libcontainer/configs/config.go#L108-L109>,
> and the passthrough is the somewhat clunky env. variable HOSTNAME set
> into the process config buffer, which the execdriver then uses
> <https://github.com/docker/docker/blob/master/daemon/execdriver/driver_unix.go#L140>
> to set up the libcontainer variant of the container's config. However, at
> least this allows for proper use via the API and with the libnetwork
> changes, gives you a more appropriate setup for differentiating hostname
> and domainname, and more importantly fixes the h ostname length max issue
> for that use case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/20200#issuecomment-182549308>.
>

thockin commented:
hmm, fat fingers.

Yes, your assessment of the `docker inspect` difference is right.

This does not fix the length issue - that is inherent to the idea that
`hostname` is an FQDN.  we could add a `--domainname` cli flag, which
would empower CLi users as much as API users, but I didn't want to do
that without deeper consideration and use-cases.  That is an easy
followup, though.

libcontainer doesn't really need to keep hostname and domainname
separate - all libcontainer cares about is what value to pass to
`syscall.Sethostname()` and that value is whatever the user passed as
`hostname` to CLI or API.

thaJeztah commented:
@thockin can you sign-off your commit? :innocent:

estesp commented:
Now that the initial overall WIP has been floated for discussion, I would suggest the libnetwork PR can be created for review as well, and those 2 file changes dropped from this PR. Don't know if @mavenugo or others want to weigh in here, or wait for a libnetwork PR. I personally think this is a reasonable fix for a long-standing issue around hostname vs. fqdn.

mavenugo commented:
thanks @thockin @estesp. yes, we will look into this mainly to make sure it doesn't introduce any serious regression.
ping @sanimej

thockin commented:
Thanks.

I will prep a "real" patch against libnetwork and sign this and update docs
in the next couple days

On Thu, Feb 11, 2016 at 9:15 AM, Madhu Venugopal <notifications@github.com>
wrote:

> thanks @thockin <https://github.com/thockin> @estesp
> <https://github.com/estesp>. yes, we will look into this mainly to make
> sure it doesn't introduce any serious regression.
> ping @sanimej <https://github.com/sanimej>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/20200#issuecomment-182962773>.
>

jhowardmsft commented:
Some missing CI jobs here. They need kicking off again. Signing the commit and force pushing should resolve that.

sanimej commented:
@thockin The change looks ok for fixing the API. But in the current form it breaks the existing behavior for net host. I think we can retain the current behavior by not splitting the os.Hostname() into host and domain name in initializeNetworking

```
        if container.HostConfig.NetworkMode.IsHost() {
                container.Config.Hostname, err = os.Hostname()
                if err != nil {
                        return err
                }

                parts := strings.SplitN(container.Config.Hostname, ".", 2) 
                if len(parts) > 1 {
                        container.Config.Hostname = parts[0]
                        container.Config.Domainname = parts[1]
                }       

        }
```

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23732/console):

~~~console

---
tainer 668fb7ae06ea
Successfully built 02008e27e8ed

---> Making bundle: validate-dco (in bundles/1.11.0-dev/validate-dco)
These commits do not have a proper 'Signed-off-by:' marker:
 - 41228686668eeb59e8618ce02b712ac58a68e41b

Please amend each commit to include a properly formatted DCO marker.

Visit the following URL for information about the Docker DCO:
 https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson3221637514991251591.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14945/console):

~~~console

---
XPERIMENTAL is set: building experimental features

---> Making bundle: validate-dco (in bundles/1.11.0-dev/validate-dco)
These commits do not have a proper 'Signed-off-by:' marker:
 - 41228686668eeb59e8618ce02b712ac58a68e41b

Please amend each commit to include a properly formatted DCO marker.

Visit the following URL for information about the Docker DCO:
 https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson6486137826857243131.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/970/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.2 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.7%
 18.3%
 22.0%
 25.7%
 29.4%
 33.0%
 36.7%
 40.4%
 44.1%
 47.7%
 51.4%
 55.1%
 58.8%
 62.4%
 66.1%
 69.8%
 73.5%
 77.1%
 80.8%
 84.5%
 88.2%
 91.9%
 95.5%
 99.2%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.40 m   54.40 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           966938169 Bytes/sec.
   Speed :           55328.645 MegaBytes/min.
   Ended : Friday, February 12, 2016 12:53:30 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-6459b8
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1856/console):

~~~console

---
/tlsconfig	[no test files]
ok  	github.com/docker/docker/pkg/truncindex	0.045s
ok  	github.com/docker/docker/pkg/urlutil	0.045s
ok  	github.com/docker/docker/pkg/useragent	0.045s
ok  	github.com/docker/docker/pkg/version	0.048s
?   	github.com/docker/docker/profiles/apparmor	[no test files]
ok  	github.com/docker/docker/profiles/seccomp	0.063s
ok  	github.com/docker/docker/reference	0.097s
ok  	github.com/docker/docker/registry	0.388s
ok  	github.com/docker/docker/runconfig	0.059s
--- FAIL: TestParseHostname (0.00s)
	parse_test.go:395: Expected the config to have 'hostname' as hostname, got 'hostname.domainname'
FAIL
FAIL	github.com/docker/docker/runconfig/opts	0.098s
ok  	github.com/docker/docker/utils	0.051s
ok  	github.com/docker/docker/volume	0.047s
ok  	github.com/docker/docker/volume/drivers	15.054s
ok  	github.com/docker/docker/volume/local	0.051s
ok  	github.com/docker/docker/volume/store	15.034s
?   	github.com/docker/docker/volume/testutils	[no test files]
Build step 'Execute shell' mark
---

~~~

thockin committed:
Don't smoosh hostname and domainname in API

This allows users to provide a FQDN as hostname or to use distinct hostname and
domainname parts.  Depends on https://github.com/docker/libnetwork/pull/950

Signed-off-by: Tim Hockin <thockin@google.com>

thockin commented:
@sanimej  re --net-host : Do you think it is sufficient to set the hostname and not the domainname?  Or should we actually get the domainname (exec `hostname -d`) and set the domainname field?

thockin commented:
Also - thoughts on what docs need to be updated?

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23763/console):

~~~console

---
c/github.com/docker/go-units".Ulimit in assignment
daemon/container_operations_unix.go:179: cannot use rl (type *"github.com/docker/go-units".Rlimit) as type *"github.com/docker/docker/vendor/src/github.com/docker/go-units".Rlimit in append
daemon/container_operations_unix.go:195: cannot use rlimits (type []*"github.com/docker/docker/vendor/src/github.com/docker/go-units".Rlimit) as type []*"github.com/docker/go-units".Rlimit in field value
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson2659967468130241024.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6215/console):

~~~console

---
c/github.com/docker/go-units".Ulimit in assignment
daemon/container_operations_unix.go:179: cannot use rl (type *"github.com/docker/go-units".Rlimit) as type *"github.com/docker/docker/vendor/src/github.com/docker/go-units".Rlimit in append
daemon/container_operations_unix.go:195: cannot use rlimits (type []*"github.com/docker/docker/vendor/src/github.com/docker/go-units".Rlimit) as type []*"github.com/docker/go-units".Rlimit in field value
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson2539252139569450376.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14977/console):

~~~console

---
c/github.com/docker/go-units".Ulimit in assignment
daemon/container_operations_unix.go:179: cannot use rl (type *"github.com/docker/go-units".Rlimit) as type *"github.com/docker/docker/vendor/src/github.com/docker/go-units".Rlimit in append
daemon/container_operations_unix.go:195: cannot use rlimits (type []*"github.com/docker/docker/vendor/src/github.com/docker/go-units".Rlimit) as type []*"github.com/docker/go-units".Rlimit in field value
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson8499060438865197331.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1887/console):

~~~console

---
perations_unix.go:179:29: error: argument 2 has incompatible type
   rlimits = append(rlimits, rl)
                             ^
daemon/container_operations_unix.go:195:33: error: incompatible type for field 13 in struct construction
   Rlimits:                      rlimits,
                                 ^
daemon/container_operations_unix.go:141:18: error: incompatible types in assignment
   ulIdx[ul.Name] = ul
                  ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson8607958572844247359.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21748/console):

~~~console

---
github.com/docker/go-units".Rlimit in append
daemon/container_operations_unix.go:195: cannot use rlimits (type []*"github.com/docker/docker/vendor/src/github.com/docker/go-units".Rlimit) as type []*"github.com/docker/go-units".Rlimit in field value
---
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson5237340480233157785.sh
+ set +x


---

~~~

sanimej commented:
@thockin For --net=host mode its sufficient to set only the hostname. This will cover the case where hostname is an fqdn. When the hostname is not an fqdn, /etc/hosts should typically have the fqdn. Since the /etc/hosts file is copied in the --net=host mode container should get the right dnsdomainname.

sanimej commented:
@thaJeztah Can you please help in identifying the docs that need to be updated for this change ? I don't see hostname & fqdn being specifically discussed in any of the docs. But probably we have to update the API documentation to clarify the behavior ?

thaJeztah commented:
@sanimej I'll have a look at the docs; from a design perspective, does this LGTY (the networking team?) if it does, we can move this forward to code review

sanimej commented:
@thaJeztah Yes, This is a reasonable solution without adding any UI or API changes.

There is one expected change in docker inspect output though; as mentioned earlier by @estesp 

>When you say docker inspect is changed, I assume you mean it will just show the entire FQDN as Hostname with a blank Domainname for the CLI usecase of --host=hostname.domainname, which would have been shown separately due to the splitting logic being at the "front-end" prior to this PR.

thaJeztah commented:
For the documentation, I think we need;

- A mention in the API changes, because specifying both `hostname` and `domainname` gives a different result; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v123-api-changes 
- Possibly update the example output of the inspect response as mentioned above

IIUC, the `--hostname` on the CLI will keep the same behavior?

thockin commented:
API doc updated.

I did not find any docs that talk about hostname/domainnname in particular - is there one you were thinking about?  

Yes, CLI behavior is unchanged.

rebased and repushed.  Still waiting on the libnetwork change.

calavera commented on daemon/container_operations_unix.go:
```diff
@@ -23,9 +23,9 @@ import (
 	"github.com/docker/docker/pkg/mount"
 	"github.com/docker/docker/pkg/stringid"
 	"github.com/docker/docker/runconfig"
+	"github.com/docker/docker/vendor/src/github.com/docker/go-units"
```
we should not change this, the old path was correct.

thockin replied:
ugh, will fix

calavera commented on runconfig/opts/parse.go:
```diff
@@ -13,11 +13,11 @@ import (
 	flag "github.com/docker/docker/pkg/mflag"
 	"github.com/docker/docker/pkg/mount"
 	"github.com/docker/docker/pkg/signal"
+	"github.com/docker/docker/vendor/src/github.com/docker/go-units"
```
we should not change this, the old path was correct.

estesp replied:
I've been having trouble with my vim go-plugins setup making this same change; only on `go-units`  Haven't figured out what is special, although one time I got frustrated enough to play with GOPATH and other bits until I got it to not change it for that session :(

thockin commented:
import paths fixed

thockin commented:
libnetwork change is in.  waiting for it to be vendored

cpuguy83 commented:
@thockin it's vendored now.

thockin commented:
Rebased and test fixed, I think

thaJeztah commented:
Thanks @thockin

thaJeztah commented:
oh cr*p, @thockin looks like it needs another rebase already :cry:

thockin commented:
rebased

On Tue, Mar 15, 2016 at 3:48 AM, Sebastiaan van Stijn <
notifications@github.com> wrote:

> oh cr*p, @thockin <https://github.com/thockin> looks like it needs
> another rebase already [image: :cry:]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/docker/docker/pull/20200#issuecomment-196762645
>

thaJeztah commented:
ping @estesp @cpuguy83 PTAL

estesp commented:
LGTM. Thanks @thockin

thockin commented:
Anything I need to do regarding the failed windows tests?

cpuguy83 commented:
Can we "fix" the inspect output for older clients?

thockin commented:
@cpuguy83 I think I am out of my depth on that question.

calavera commented:
LGTM, let's move this forward to docs review.

thaJeztah commented:
thanks, LGTM - not sure if there's anything else to document :smile:

calavera commented:
Docs LGTM too.
